File: overmapper-2.rb
#!/usr/bin/env ruby
# coding: utf-8

require 'oj'

module OverMapper
  class Error < StandardError; end

  class Converter
    LINES = {
      4194424 => "\u2502",
      4194417 => "\u2500",
      4194413 => "\u2514",
      4194412 => "\u250C",
      4194411 => "\u2510",
      4194410 => "\u2518",
      4194420 => "\u251C",
      4194422 => "\u2534",
      4194421 => "\u2524",
      4194423 => "\u252C",
      4194414 => "\u253C",
    }
    ROADS = {
      'north' => "│",
      'south' => "│",
      'west' => "─",
      'east' => "─",
      'nesw' => "┼",
      'ns' => "│",
      'ew' => "─",
      'wn' => "┘",
      'ne' => "└",
      'sw' => "┐",
      'es' => "┌",
      'esw' => "┬",
      'nsw' => "┤",
      'new' => "┴",
      'nes' => "├",
    }
    DIRS = %w{north east south west ns ew sn we}
    HTML_START = """<html><head><style>
body { background: black; color: #aaaaaa; }
.cl_white { color: #ffffff; }
.cl_blue { color: #0000ff; }
.cl_red { color: #ff0000; }
.cl_brown { color: #a52a2a; }
.cl_green { color: #008000; }
.cl_cyan { color: #00ffff; }
.cl_dark_gray { color: #a9a9a9; }
.cl_magenta { color: #ff00ff; }
.cl_yellow { color: #ffff00; }
.cl_light_blue { color: #add8e6; }
.cl_light_green { color: #90ee90; }
.cl_light_red { color: #ff5555; }
.cl_i_ltred { color: black; background: #ff5555; }
.cl_light_gray { color: #d3d3d3; }
.cl_i_ltgray { color: black; background: #d3d3d3; }
.cl_light_cyan { color: #e0ffff; }
.cl_ltgray_yellow { color: #d3d3d3; background: #ffff00; }
.cl_pink { color: #ffc0cb; }
.cl_yellow_magenta { color: #ffff00; background: #ff00ff; }
.cl_white_magenta { color: #ffffff; background: #ff00ff; }
.cl_i_magenta { color: black; background: #ff00ff; }
.cl_pink_magenta { color: #ffc0cb; background: #ff00ff;  }
.cl_i_green { color: black; background: #008000; }
.cl_i_brown { color: black; background: #a52a2a; }
.cl_h_yellow { color: #ffff00; background: #0000ff; }
.cl_h_dkgray { color: #a9a9a9; background: #0000ff; }
.cl_i_ltblue { color: black; background: #add8e6; }
.cl_i_blue { color: black; background: #0000ff; }
.cl_i_red { color: black; background: #ff0000; }
.cl_ltgreen_yellow { color: #d3d3d3; background: #ffff00; }
.cl_white_white { background: white; }
.cl_i_ltcyan { color: black; background: #e0ffff; }
.cl_yellow_cyan { color: #ffff00; background: #00ffff; }
</style></head><body><pre>"""
    HTML_END = """</pre></body></html>"""
    FORMATS = %i{txt html}

    attr_accessor :data, :path

    def initialize(data_path, logger: nil)
      @logger = logger
      @data   = nil
      @path   = File.join(data_path, 'terrain.dat')

      if File.exists? @path
        load_data(@path)
      else
        log :warn, 'No data file found'
      end
    end

    def preprocess(cdda_path)
      in_paths = ([File.join(cdda_path, 'data/json/overmap_terrain.json')] + \
                  Dir.glob(File.join(cdda_path, 'data/json/mapgen/*.json')))

      log :debug, 'Preprocessing terrain data'
      @data = Hash.new
      in_paths.each do |p|
        Oj.load_file(p).each do |e|
          next if (e['id'].nil? || e['color'].nil? || e['sym'].nil?)

          begin
            id    = e['id']
            color = e['color']
            syms  = process_sym(e['sym'])
            tile  = {syms: syms, color: color}

            if @data.has_key? id
              log :warn, "Duplicate terrain id #{id}"
            end

            @data[id] = tile
          rescue StandardError => error
            log :error, e.pretty_inspect
            log :error, error.pretty_inspect
          end
        end
      end
      log :info, 'Saving terrain data'
      File.open(@path, 'wb') {|f| f.write(Marshal.dump(@data)) }
    end

    def load_data(dat_path)
      @data = File.open(dat_path, 'rb') {|f| Marshal.load(f) }
    end

    def convert(dir_path, format: :html, width: 180, height: 180, layer: 10)
      unless @data
        raise Error, 'No terrain.dat loaded'
      end
      unless FORMATS.include? format
        raise ArgumentError, "Unknown format: #{format}"
      end

      empty_line = ' ' * width
      maps = get_maps(dir_path).collect do |row|
        row.collect do |map|
          unless map
            next Array.new(height, empty_line)
          end

          data = get_layer(map, layer)
          strings = Array.new
          line = String.new
          pos = 0

          data.each do |id, len|
            sym, color = get_tile_data(id)
            if format == :html
              # does not care about proper Unicode glyphs
              if sym == '<'
                sym = '&#x3c;'
              elsif sym == '>'
                sym = '&#x3e;'
              end
            end
            while len > 0
              line_left = width - pos
              how_many = [line_left, len].min
              if format == :html && color
                line += "<span class=\"cl_#{color}\">"
              end
              line += sym * how_many
              if format == :html && color
                line += '</span>'
              end
              pos += how_many
              len -= how_many
              if pos == width
                strings.push(line)
                line = String.new
                pos = 0
              end
            end
          end

          strings
        end.transpose.collect(&:join)
      end

      if format == :html
        puts HTML_START
      end
      maps.each {|m| puts m }
      if format == :html
        puts HTML_END
      end
    end

    private

    def get_tile_data(id)
      tile = @data[id]
      if tile
        return [tile[:syms].first, tile[:color]]
      else
        parts = id.split('_')
        suffix = parts.pop
        nid = parts.join('_')

        if nid == 'road'
          sym = ROADS[suffix]
          unless sym
            log :warn, "No road for suffix #{suffix}"
            return ['?', nil]
          end
        else
          tile = @data[nid]
          unless tile
            log :warn, "Tile not found #{id} (#{nid} #{suffix})"
            return ['?', nil]
          end
          dir = DIRS.index(suffix) % 4 rescue 0
          sym = tile[:syms][dir] || tile[:syms].first
          color = tile[:color]
        end
      end
      [sym, color]
    end

    def get_layer(json_path, layer)
      raw = File.read(json_path).lines
      header = raw.shift
      unless header.match(/# version 25/)
        raise Error, "Unknown header: #{header}"
      end

      Oj::Doc.open(raw.join("\n")) {|d| d.fetch("/layers/#{layer + 1}") }
    end

    def get_maps(dir_path)
      paths = Dir.glob(File.join(dir_path, 'o.*'))
      if paths.empty?
        raise Error, 'No maps found'
      end

      coords = Array.new
      seen   = Hash.new
      paths.each do |p|
        x, y = p.match(/o.(-?\d+).(-?\d+)/).captures.map(&:to_i)
        coords.push([x, y])
        seen["#{x} #{y}"] = p
      end

      coords = coords.transpose
      min_x, max_x = coords.first.minmax
      min_y, max_y = coords.last.minmax

      min_y.upto(max_y).collect do |y|
        min_x.upto(max_x).collect do |x|
          seen["#{x} #{y}"]
        end
      end
    end

    def process_sym(raw)
      if raw.nil?
        log :warn, 'No raw symbol given'
        return ['?']
      end

      [raw].flatten.collect do |r|
        if r <= 255
          r.chr
        else
          if LINES.has_key? r
            LINES[r]
          else
            log :warn, 'Raw symbol not found in LINES'
            '?'
          end
        end
      end
    end

    def log(level, msg)
      if @logger
        @logger.send(level, msg)
      end
    end
  end
end

# provide a really dumb user interface
if __FILE__ == $0
  def usage
    STDERR.puts """Usage:
      #{$0} preprocess /path/to/cdda_dir

      Convert CDDA overmap terrain data for later use.
      File 'terrain.dat' will be created in current directory.
      NOTE: You need to have that file for convert to work.

      #{$0} convert [format] /path/to/save_dir
      
      format: html, txt. default: html
      
      This will try to convert overmap data to given format.
      It will output to stdout so you better redirect it to a file."""
    exit(2)
  end

  cmd, *args = ARGV
  unless cmd == 'preprocess' || cmd == 'convert'
    usage
  end

  require 'logger'
  logger = Logger.new(STDERR)
  logger.level = Logger::DEBUG
  logger.formatter = lambda {|s, d, p, m| "#{d} #{s} -- #{m}\n" }

  cn = OverMapper::Converter.new(File.absolute_path('./'), logger: logger)
  if cmd == 'preprocess'
    cn.preprocess(args.first)
  else
    format = :html
    if args.length > 1
      unless ['html', 'txt'].include? args.first
        usage
      end
      format = args.shift.to_sym
    end
    cn.convert(File.absolute_path(args.first), format: format)
  end
end

File: overmapper-2.go
// See LICENSE.txt for licensing information.

package main

import (
	"bytes"
	"encoding/gob"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

// Terrain holds possible symbols (for different orientations) and color name
// string for an overmap terrain entity.
type Terrain struct {
	Symbols []string
	Color   string
}

var (
	flagVerbose bool // be verbose
	flagPlain   bool // convert to plain-text
	flagWidth   int  // overmap width
	flagHeight  int  // overmap height
	flagLayer   int  // overmap layer

	termap = make(map[string]Terrain, 1024) // global overmap terrain map
	lines  = map[int32]string{              // mappings for line glyphs
		4194424: "\u2502",
		4194417: "\u2500",
		4194413: "\u2514",
		4194412: "\u250C",
		4194411: "\u2510",
		4194410: "\u2518",
		4194420: "\u251C",
		4194422: "\u2534",
		4194421: "\u2524",
		4194423: "\u252C",
		4194414: "\u253C",
	}
	roads = map[string]string{ // mappings for road glyphs
		"north": "│",
		"south": "│",
		"west":  "─",
		"east":  "─",
		"nesw":  "┼",
		"esw":   "┬",
		"nsw":   "┤",
		"new":   "┴",
		"nes":   "├",
		"ns":    "│",
		"ew":    "─",
		"wn":    "┘",
		"ne":    "└",
		"sw":    "┐",
		"es":    "┌",
	}
	dirs      = []string{"north", "east", "south", "west", "ns", "ew", "sn", "we"} // possible orientation etc. strings
	htmlStart = `<html><head><style>
body { background: black; color: #aaaaaa; }
.cl_white { color: #ffffff; }
.cl_blue { color: #0000ff; }
.cl_red { color: #ff0000; }
.cl_brown { color: #a52a2a; }
.cl_green { color: #008000; }
.cl_cyan { color: #00ffff; }
.cl_dark_gray { color: #a9a9a9; }
.cl_magenta { color: #ff00ff; }
.cl_yellow { color: #ffff00; }
.cl_light_blue { color: #add8e6; }
.cl_light_green { color: #90ee90; }
.cl_light_red { color: #ff5555; }
.cl_i_ltred { color: black; background: #ff5555; }
.cl_light_gray { color: #d3d3d3; }
.cl_i_ltgray { color: black; background: #d3d3d3; }
.cl_light_cyan { color: #e0ffff; }
.cl_ltgray_yellow { color: #d3d3d3; background: #ffff00; }
.cl_pink { color: #ffc0cb; }
.cl_yellow_magenta { color: #ffff00; background: #ff00ff; }
.cl_white_magenta { color: #ffffff; background: #ff00ff; }
.cl_i_magenta { color: black; background: #ff00ff; }
.cl_pink_magenta { color: #ffc0cb; background: #ff00ff;  }
.cl_i_green { color: black; background: #008000; }
.cl_i_brown { color: black; background: #a52a2a; }
.cl_h_yellow { color: #ffff00; background: #0000ff; }
.cl_h_dkgray { color: #a9a9a9; background: #0000ff; }
.cl_i_ltblue { color: black; background: #add8e6; }
.cl_i_blue { color: black; background: #0000ff; }
.cl_i_red { color: black; background: #ff0000; }
.cl_ltgreen_yellow { color: #d3d3d3; background: #ffff00; }
.cl_white_white { background: white; }
.cl_i_ltcyan { color: black; background: #e0ffff; }
.cl_yellow_cyan { color: #ffff00; background: #00ffff; }
</style><meta content='text/html; charset=utf-8' http-equiv='Content-Type' />
</head><body><pre>`
	htmlEnd = `</pre></body></html>`
)

func init() {
	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "Usage: %s [options...] <command> <path>\n", os.Args[0])
		fmt.Fprintf(os.Stderr, "\n	Options:\n\n")
		flag.PrintDefaults()
		fmt.Fprintln(os.Stderr, `
	Commands:

	prepare /path/to/cdda

	Convert C:DDA overmap terrain data to internal format.
	File 'terrain.dat' will be created in current directory.
	NOTE: You need to have the above file for convert to work.
	      You should also re-run it periodically to update the data
	      (e.g. if you see missing tile messages when converting).

	convert /path/to/save > /path/to/output.file

	This will try to convert your save data to HTML by default.
	It will output to stdout so you most likely want to redirect it
	to a file ("> test.html"). You can change the format using the "-p"
	option (for plain text).`)
		os.Exit(1)
	}
	flag.BoolVar(&flagVerbose, "v", false, "be verbose")
	flag.BoolVar(&flagPlain, "p", false, "convert to plain-text")
	flag.IntVar(&flagWidth, "ow", 180, "overmap width")
	flag.IntVar(&flagHeight, "oh", 180, "overmap height")
	flag.IntVar(&flagLayer, "ol", 10, "overmap layer")
}

func main() {
	flag.Parse()
	if flag.NArg() != 2 {
		flag.Usage()
	}
	switch flag.Arg(0) {
	case "prepare":
		runPrepare(flag.Arg(1))
		if flagVerbose {
			fmt.Fprintf(os.Stderr, "Parsed %d terrain items\n", len(termap))
		}
		saveTermap()
	case "convert":
		loadTermap()
		runConvert(flag.Arg(1))
	default:
		flag.Usage()
	}
}

// die prints the error to stderr end exits.
func die(err error) {
	fmt.Fprintln(os.Stderr, "FATAL: ", err)
	os.Exit(2)
}

// warn prints a warning to stderr.
func warn(str string, args ...interface{}) {
	fmt.Fprintf(os.Stderr, "WARN: "+str, args...)
}

// oMapElem holds overmap data from a save.
type oMapElem struct {
	ID  string
	Len int
}

// Custom unmarshaling due to the structure of the overmap data in a save.
func (e *oMapElem) UnmarshalJSON(b []byte) (err error) {
	var id string
	var length int
	tgt := []interface{}{&id, &length}
	if err := json.Unmarshal(b, &tgt); err != nil {
		return err
	}
	e.ID = id
	e.Len = length
	return
}

// getLayer returns a parsed overmap layer or dies trying.
func getLayer(which int, path string) (layer []oMapElem, err error) {
	which++ // as we start from 1 due to silly delimiter counting later on
	file, err := os.Open(path)
	if err != nil {
		return
	}
	defer file.Close()
	if _, err = file.Seek(12, os.SEEK_CUR); err != nil {
		return
	}
	dec := json.NewDecoder(file)
	current := 0
	level := 0
	for {
		t, err := dec.Token()
		if err != nil {
			die(err)
		}
		d, ok := t.(json.Delim)
		if !ok {
			continue
		}
		if d == '[' {
			level++
		} else if d == ']' {
			level--
		}
		if level == 1 {
			current++
		}
		if current == which {
			break
		}
	}
	err = dec.Decode(&layer)
	return
}

// runConvert is the main entry point for conversion.
// It will die on fatal errors.
func runConvert(root string) {
	maps, width, err := getMaps(root)
	if err != nil {
		die(err)
	}
	rowLineBufs := make([][]bytes.Buffer, width)
	for x := 0; x < width; x++ {
		rowLineBufs[x] = make([]bytes.Buffer, flagHeight)
	}
	if !flagPlain {
		fmt.Println(htmlStart)
	}
	for _, row := range maps {
		for x, mapPath := range row {
			if mapPath == "" {
				for line := 0; line < flagHeight; line++ {
					for i := 0; i < flagWidth; i++ {
						rowLineBufs[x][line].Write([]byte(" "))
					}
				}
				continue
			}
			data, err := getLayer(flagLayer, mapPath)
			if err != nil {
				die(err)
			}
			line := 0
			pos := 0
			for _, e := range data {
				color, symbol := getTer(e.ID)
				if !flagPlain {
					if symbol == "<" {
						symbol = "&#x3c;"
					} else if symbol == ">" {
						symbol = "&#x3e;"
					}
				}
				for e.Len > 0 {
					lineLeft := flagWidth - pos
					var howMany int
					if lineLeft < e.Len {
						howMany = lineLeft
					} else {
						howMany = e.Len
					}
					if color != "" && !flagPlain {
						rowLineBufs[x][line].WriteString("<span class=\"cl_")
						rowLineBufs[x][line].WriteString(color)
						rowLineBufs[x][line].WriteString("\">")
					}
					for c := 0; c < howMany; c++ {
						rowLineBufs[x][line].WriteString(symbol)
					}
					if color != "" && !flagPlain {
						rowLineBufs[x][line].WriteString("</span>")
					}
					pos += howMany
					e.Len -= howMany
					if pos == flagWidth {
						pos = 0
						line++
					}
				}
			}
		}
		for line := 0; line < flagHeight; line++ {
			for _, bufs := range rowLineBufs {
				bufs[line].WriteTo(os.Stdout)
				bufs[line].Reset()
			}
			fmt.Printf("\n")
		}
	}
	if !flagPlain {
		fmt.Println(htmlEnd)
	}
}

// getTer tries to retrieve color and symbol strings for a given overmap
// terrain id. It handles rotations, lines and roads (as special cases).
func getTer(id string) (color, symbol string) {
	tile, ok := termap[id]
	if ok {
		return tile.Color, tile.Symbols[0]
	}
	splitAt := strings.LastIndex(id, "_")
	if splitAt < 1 {
		warn("Couldn't split id: %s\n", id)
		return "", "?"
	}
	suffix := id[splitAt+1:]
	nid := id[0:splitAt]
	if nid == "road" {
		symbol, ok = roads[suffix]
		if !ok {
			warn("No road suffix for %s (%s)\n", suffix, id)
			return "", "?"
		}
		return
	}
	tile, ok = termap[nid]
	if !ok {
		warn("Tile not found %s (%s %s)\n", nid, id, suffix)
		return "", "?"
	}
	dir := 0
	for i, d := range dirs {
		if d == suffix {
			dir = i
			break
		}
	}
	dir = dir % 4
	if dir < len(tile.Symbols) {
		symbol = tile.Symbols[dir]
	} else {
		symbol = tile.Symbols[0]
	}
	color = tile.Color
	return
}

// getMaps looks over a save directory and composes an array of arrays
// of maps for processing. The returned array has dimensions of the extent
// of the overmaps saved, with empty overmaps being empty strings.
// It also returns the width in overmaps as it's needed to allocate output
// buffers later.
func getMaps(root string) (maps [][]string, width int, err error) {
	type coord struct {
		x, y int
	}
	var min_x, max_x int
	var min_y, max_y int
	seen := make(map[coord]string)
	err = filepath.Walk(root,
		func(path string, info os.FileInfo, ierr error) (oerr error) {
			if ierr != nil {
				return ierr
			}
			if info.IsDir() {
				return nil
			}
			fname := filepath.Base(path)
			if ok, _ := filepath.Match("o.*", fname); !ok {
				return nil
			}
			if flagVerbose {
				fmt.Fprintf(os.Stderr, "File: %s\n", path)
			}
			parts := strings.Split(fname, ".")
			if len(parts) != 3 {
				warn("Malformed file match: %s\n", fname)
				return nil
			}
			x, oerr := strconv.Atoi(parts[1])
			if oerr != nil {
				return
			}
			y, oerr := strconv.Atoi(parts[2])
			if oerr != nil {
				return
			}
			seen[coord{x, y}] = path
			if x < min_x {
				min_x = x
			}
			if x > max_x {
				max_x = x
			}
			if y < min_y {
				min_y = y
			}
			if y > max_y {
				max_y = y
			}
			return nil
		})
	width = -min_x + 1 + max_x
	height := -min_y + 1 + max_y
	if flagVerbose {
		fmt.Fprintf(os.Stderr, "Maps range from %dx%d to %dx%d for an area of %dx%d\n",
			min_x, min_y, max_x, max_y, width, height)
	}
	maps = make([][]string, height)
	for y := 0; y < height; y++ {
		maps[y] = make([]string, width)
		for x := 0; x < width; x++ {
			rx := min_x + x
			ry := min_y + y
			if path, ok := seen[coord{rx, ry}]; ok {
				maps[y][x] = path
			}
		}
	}
	return
}

// runPrepare is the main entry point for data preprocessing.
// It will die on fatal errors.
// The idea here is that we walk all the possible JSONs paths and try to
// extract anything that looks like a valid overmap terrain entity.
// We can then gob-encode what we found and have it ready to use for
// all future converts.
func runPrepare(root string) {
	err := processTerrainData(filepath.Join(root, "data", "json",
		"overmap_terrain.json"))
	if err != nil {
		die(err)
	}
	err = filepath.Walk(filepath.Join(root, "data", "json", "mapgen"), walkForMapgen)
	if err != nil {
		die(err)
	}
	err = filepath.Walk(filepath.Join(root, "data", "mods"), walkForMods)
	if err != nil {
		die(err)
	}
}

// walkForMapgen walks recursively looking for JSON files and tries to extract
// overmap terrain entities form them.
func walkForMapgen(path string, info os.FileInfo, ierr error) (oerr error) {
	if ierr != nil {
		return ierr
	}
	if info.IsDir() {
		return nil
	}
	if filepath.Ext(path) != ".json" {
		return nil
	}
	return processTerrainData(path)
}

// walkForMods walks recursively looking for JSON files and tries to extract
// overmap terrain entities form them.
func walkForMods(path string, info os.FileInfo, ierr error) (oerr error) {
	if ierr != nil {
		return ierr
	}
	if info.IsDir() {
		return nil
	}
	if filepath.Base(path) != "overmap_terrain.json" {
		return nil
	}
	return processTerrainData(path)
}

// processTerrainData tries to extract all overmap entities form a given
// JSON file.
func processTerrainData(path string) (err error) {
	type oter struct {
		ID    string      `json:"id"`
		Color string      `json:"color"`
		Syms  interface{} `json:"sym"`
	}
	if flagVerbose {
		fmt.Fprintf(os.Stderr, "File: %s\n", path)
	}
	file, err := os.Open(path)
	if err != nil {
		return
	}
	defer file.Close()
	dec := json.NewDecoder(file)
	_, err = dec.Token() // Skip over '['
	if err != nil {
		return
	}
	var rawTer oter
	for dec.More() {
		err = dec.Decode(&rawTer)
		if err != nil {
			warn("Bad item format: %s\n", err)
			continue
		}
		if rawTer.Syms == nil {
			continue
		}
		if _, ok := termap[rawTer.ID]; ok {
			warn("Terrain %v already defined\n", rawTer.ID)
		}
		var syms []string
		switch tsym := rawTer.Syms.(type) {
		case float64:
			syms = make([]string, 1)
			syms[0] = parseSym(int32(tsym))
		case []interface{}:
			syms = make([]string, len(tsym))
			for i, ttsym := range tsym {
				// below will blow up with appropriate message on wrong type
				syms[i] = parseSym(int32(ttsym.(float64)))
			}
		default:
			warn("Unkown sym type: %v\n", tsym)
			continue
		}
		ter := Terrain{syms, rawTer.Color}
		termap[rawTer.ID] = ter
		if flagVerbose {
			fmt.Fprintf(os.Stderr, "Parsed terrain: %v = %v\n", rawTer.ID, ter)
		}
	}
	return
}

// parseSym tries to parse a symbol (integer) into a string.
func parseSym(raw int32) string {
	if raw <= 255 {
		return string(raw)
	}
	if str, ok := lines[raw]; ok {
		return str
	}
	warn("No LINE found for: %v\n", raw)
	return "?"
}

// saveTermap saves the global terrain data or dies.
func saveTermap() {
	file, err := os.Create("terrain.dat")
	if err != nil {
		die(err)
	}
	defer file.Close()
	enc := gob.NewEncoder(file)
	if err = enc.Encode(termap); err != nil {
		die(err)
	}
	if flagVerbose {
		fmt.Fprintf(os.Stderr, "Saved terrain map to terrain.dat\n")
	}
}

// loadTermap loads the global terrain data or dies.
func loadTermap() {
	file, err := os.Open("terrain.dat")
	if err != nil {
		die(err)
	}
	defer file.Close()
	dec := gob.NewDecoder(file)
	if err = dec.Decode(&termap); err != nil {
		die(err)
	}
	if flagVerbose {
		fmt.Fprintf(os.Stderr, "Loaded terrain map (%d items)\n", len(termap))
	}
}
